mmcv/ops/csrc/common/cuda/rans_cuda_kernel.cuh

```diff
@@ -0,0 +1,206 @@
+                            const std::vector<RansCUDAKernelResult>& results,
+                            const int begin_idx, const int end_idx,
+                            std::vector<uint32_t>& output, uint32_t& nbyte) {
+  Rans64State rans;
+  Rans64EncInit(&rans);
+
+  uint32_t* ptr = output.data() + output.size();
+  assert(ptr != nullptr);
+
+  for (int i = end_idx - 1; i >= begin_idx; i--) {
+    if (results[i].bypass_mode) {
+      // unlikely...
+      /* Determine the number of bypasses (in bypass_precision size) needed to
+       * encode the raw value. */
+      std::vector<RansSymbol> _syms;
+      int32_t n_bypass = 0;
+      uint32_t raw_val = results[i].raw_value;
+      while ((raw_val >> (n_bypass * bypass_precision)) != 0) {
+        ++n_bypass;
+      }
+
+      /* Encode number of bypasses */
+      int32_t val = n_bypass;
+      while (val >= max_bypass_val) {
+        _syms.push_back({max_bypass_val, max_bypass_val + 1, true});
+        val -= max_bypass_val;
+      }
+      _syms.push_back(
+          {static_cast<uint16_t>(val), static_cast<uint16_t>(val + 1), true});
+
+      /* Encode raw value */
+      for (int32_t j = 0; j < n_bypass; ++j) {
+        const int32_t val =
+            (raw_val >> (j * bypass_precision)) & max_bypass_val;
+        _syms.push_back(
+            {static_cast<uint16_t>(val), static_cast<uint16_t>(val + 1), true});
+      }
+
+      while (!_syms.empty()) {
+        const RansSymbol sym = _syms.back();
+        Rans64EncPutBits(&rans, &ptr, sym.start, bypass_precision);
+        _syms.pop_back();
+      }
+    }
+    Rans64EncPut(&rans, &ptr, symbols[i].start, symbols[i].range, precision);
+  }
+
+  Rans64EncFlush(&rans, &ptr);
+
+  nbyte = std::distance(ptr, output.data() + output.size()) * sizeof(uint32_t);
+}
+
+// ------------------- PMF TO QUANTIZED CDF -------------------
+
+__global__ void pmf_to_quantized_cdf_cuda_kernel(
+    const float* pmfs, const int* pmf_lengths, const float* tail_masses,
+    int* quantized_cdfs, const int pmfs_size0, const int pmfs_size1,
+    const int cdf_max_length) {
+  const int i = blockIdx.x;
+  const int j = threadIdx.x;
+  const int pmf_length = pmf_lengths[i];
+  if (i >= pmfs_size0 || j >= pmf_length + 2) return;
+
+  const float* pmf = pmfs + i * pmfs_size1;
+  const float tail_mass = tail_masses[i];
+  __shared__ int shared_cdf[THREADS_PER_BLOCK];
+
+  if (j == 0) {
+    shared_cdf[j] = 0;
+  } else if (j == pmf_length + 1) {
+    shared_cdf[j] = round(tail_mass * (1 << precision)); /* freq of tail mass */
+  } else {
+    shared_cdf[j] = round(pmf[j - 1] * (1 << precision));
+  }
+
+  __shared__ uint32_t total;
+  if (j == 0) {
+    total = 0;
+  }
+  __syncthreads();
+  atomicAdd(&total, shared_cdf[j]);
+  __syncthreads();
+  assert(total != 0);
+
+  shared_cdf[j] =
+      (static_cast<uint64_t>(1 << precision) * ((uint32_t)shared_cdf[j])) /
+      total;
+  __syncthreads();
+
+  if (j == 0) {
```

Comment:
could use [block scan](https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda)

Reply:
While this operation has some similarities to the order-free sum computation performed by block scan, I think it cannot be directly optimized using block scan because it is not order-free. Each `shared_cdf[k]` is dependent on the previous sum value. Furthermore, this function is used infrequently and is only used once during the network initialization.

Reply:
Thanks, I will keep it now and add a //TODO
```cuh
  // TODO: this could be accelerated by using a block scan
```